File: src/harness/a2ui/composer.py
"""Composer: convierte el *plan de UI* que emite el LLM en envelopes A2UI válidos.

Decisión de arquitectura (ver documentacion/adr/0002):
el modelo NO emite protocolo crudo. Emite un plan compacto; Python lo valida
contra el catálogo, resuelve bindings y lo traduce a A2UI v0.9.1. Así:
  - una alucinación de componente nunca rompe el render (se poda + se repara),
  - el protocolo puede migrar (v0.9.1 -> v1.0) sin tocar el prompt,
  - el composer es determinista y testeable sin llamar al modelo.

Plan esperado (JSON):
{
  "title": "Reestructura tu saldo",
  "summary": "una línea de texto para el chat",
  "root": "root",
  "data": {"plan": {"selected": "12m"}},
  "components": [
     {"id": "root", "component": "Column", "props": {"children": ["kpi", "opts"]}},
     ...
  ]
}
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .catalog import COMPONENTS
from .messages import create_surface, update_components, update_data_model


@dataclass
class CompileResult:
    ok: bool
    messages: list[dict] = field(default_factory=list)
    components: list[dict] = field(default_factory=list)
    data: dict = field(default_factory=dict)
    errors: list[str] = field(default_factory=list)
    dropped: list[str] = field(default_factory=list)


def _is_binding(value: Any) -> bool:
    return isinstance(value, dict) and set(value.keys()) <= {"path"} and "path" in value


def _coerce_prop(comp: str, cid: str, pname: str, spec: dict, value: Any, errors: list[str]) -> Any:
    t = spec["type"]
    if t == "binding":
        if _is_binding(value):
            return value
        return value  # literal: string/number/bool también es válido
    if t == "enum":
        if value in spec["values"]:
            return value
        errors.append(f"{cid}({comp}).{pname}='{value}' no está en {spec['values']}")
        return spec.get("default", spec["values"][0])
    if t == "number":
        try:
            return float(value) if isinstance(value, str) else value
        except (TypeError, ValueError):
            errors.append(f"{cid}({comp}).{pname} debe ser número")
            return spec.get("default", 0)
    if t == "string":
        return value if isinstance(value, str) else str(value)
    if t == "boolean":
        return bool(value)
    if t in ("componentId", "componentIds", "objectList", "action"):
        return value
    return value


def validate_components(raw: list[dict]) -> tuple[list[dict], list[str], list[str]]:
    """Valida contra el catálogo. Devuelve (componentes_ok, errores, ids_descartados)."""
    errors: list[str] = []
    dropped: list[str] = []
    by_id: dict[str, dict] = {}

    for item in raw or []:
        if not isinstance(item, dict):
            errors.append("componente que no es objeto")
            continue
        cid = item.get("id")
        comp = item.get("component")
        props = item.get("props") or {k: v for k, v in item.items() if k not in ("id", "component")}
        if not cid or not comp:
            errors.append(f"componente sin id/component: {str(item)[:80]}")
            continue
        if comp not in COMPONENTS:
            errors.append(f"'{comp}' no existe en el catálogo (id={cid})")
            dropped.append(cid)
            continue
        spec = COMPONENTS[comp]["props"]
        clean: dict[str, Any] = {}
        for pname, pspec in spec.items():
            if pname in props and props[pname] is not None:
                clean[pname] = _coerce_prop(comp, cid, pname, pspec, props[pname], errors)
            elif "default" in pspec:
                clean[pname] = pspec["default"]
            elif pspec.get("required"):
                errors.append(f"{cid}({comp}) falta prop requerida '{pname}'")
        for extra in set(props) - set(spec):
            errors.append(f"{cid}({comp}) prop desconocida '{extra}' (ignorada)")
        by_id[cid] = {"id": cid, "component": comp, **clean}

    # Poda de referencias rotas: children/child que apuntan a ids inexistentes.
    for cid, node in list(by_id.items()):
        for key in ("children", "child"):
            if key not in node:
                continue
            if key == "children" and isinstance(node[key], list):
                alive = [c for c in node[key] if c in by_id]
                if len(alive) != len(node[key]):
                    errors.append(f"{cid}: referencias rotas podadas en children")
                node[key] = alive
            elif key == "child" and node[key] not in by_id:
                errors.append(f"{cid}: child '{node[key]}' no existe")
                node.pop(key)
                dropped.append(cid)
                by_id.pop(cid, None)

    return list(by_id.values()), errors, dropped


def compile_plan(plan: dict, surface_id: str, first_render: bool) -> CompileResult:
    components, errors, dropped = validate_components(plan.get("components", []))
    ids = {c["id"] for c in components}
    root = plan.get("root") or (components[0]["id"] if components else None)

    if not components or root not in ids:
        errors.append("plan sin root renderizable")
        return CompileResult(ok=False, errors=errors, dropped=dropped)

    # A2UI v0.9: el root es el primer componente de la lista.
    components.sort(key=lambda c: 0 if c["id"] == root else 1)

    msgs: list[dict] = []
    if first_render:
        msgs.append(create_surface(surface_id))
    data = plan.get("data") or {}
    if isinstance(data, dict) and data:
        msgs.append(update_data_model(surface_id, "/", data))
    msgs.append(update_components(surface_id, components))

    return CompileResult(True, msgs, components, data if isinstance(data, dict) else {}, errors, dropped)


def fallback_plan(title: str, body: str, retry_action: str = "reintentar") -> dict:
    """Superficie mínima garantizada: la demo nunca se queda en blanco."""
    return {
        "title": title,
        "summary": body,
        "root": "root",
        "data": {},
        "components": [
            {"id": "root", "component": "Column", "props": {"children": ["t", "b", "cta"], "gap": "md"}},
            {"id": "t", "component": "Text", "props": {"text": title, "variant": "h2"}},
            {"id": "b", "component": "Text", "props": {"text": body, "variant": "body"}},
            {
                "id": "cta",
                "component": "ActionButton",
                "props": {
                    "text": "Reintentar",
                    "variant": "secondary",
                    "action": {"event": {"name": retry_action, "params": {}}},
                },
            },
        ],
    }

File: web/src/agent/composer.ts
/**
 * Puerto de `a2ui/composer.py` + `a2ui/messages.py`.
 *
 * El modelo NO emite protocolo crudo: emite un plan compacto que aquí se
 * valida contra el catálogo y se traduce a envelopes A2UI v0.9.1. Esa
 * validación es la razón por la que una alucinación de componente no rompe
 * el render — se poda, se reporta y, si el plan queda inservible, se repara
 * o cae al fallback.
 */
import type { CatalogDocument, PropSpec } from "./catalog";

const A2UI_VERSION = "v0.9.1";

export interface PlanComponent {
  id?: unknown;
  component?: unknown;
  props?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface UiPlan {
  title?: string;
  summary?: string;
  root?: string;
  data?: Record<string, unknown>;
  components?: PlanComponent[];
}

export interface CompileResult {
  ok: boolean;
  messages: Record<string, unknown>[];
  components: Record<string, unknown>[];
  data: Record<string, unknown>;
  errors: string[];
  dropped: string[];
}

// ── envelopes ────────────────────────────────────────────────────────────

function createSurface(surfaceId: string, catalogId: string): Record<string, unknown> {
  return {
    version: A2UI_VERSION,
    createSurface: { surfaceId, catalogId, sendDataModel: true },
  };
}

function updateComponents(
  surfaceId: string,
  components: Record<string, unknown>[]
): Record<string, unknown> {
  return { version: A2UI_VERSION, updateComponents: { surfaceId, components } };
}

function updateDataModel(
  surfaceId: string,
  path: string,
  value: unknown
): Record<string, unknown> {
  return { version: A2UI_VERSION, updateDataModel: { surfaceId, path, value } };
}

// ── validación ───────────────────────────────────────────────────────────

function isBinding(value: unknown): boolean {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const keys = Object.keys(value as Record<string, unknown>);
  return keys.length === 1 && keys[0] === "path";
}

function coerceProp(
  comp: string,
  cid: string,
  pname: string,
  spec: PropSpec,
  value: unknown,
  errors: string[]
): unknown {
  switch (spec.type) {
    case "binding":
      // literal (string/number/bool) o {"path": "/ptr"} — las dos son válidas
      return isBinding(value) ? value : value;
    case "enum": {
      const values = spec.values ?? [];
      if (values.includes(value as string)) return value;
      errors.push(`${cid}(${comp}).${pname}='${String(value)}' no está en ${JSON.stringify(values)}`);
      return "default" in spec ? spec.default : values[0];
    }
    case "number": {
      if (typeof value === "number") return value;
      if (typeof value === "string") {
        const n = Number(value);
        if (Number.isFinite(n)) return n;
      }
      errors.push(`${cid}(${comp}).${pname} debe ser número`);
      return "default" in spec ? spec.default : 0;
    }
    case "string":
      return typeof value === "string" ? value : String(value);
    case "boolean":
      return Boolean(value);
    default:
      // componentId | componentIds | objectList | action pasan tal cual
      return value;
  }
}

export function validateComponents(
  raw: PlanComponent[] | undefined,
  catalog: CatalogDocument
): { components: Record<string, unknown>[]; errors: string[]; dropped: string[] } {
  const errors: string[] = [];
  const dropped: string[] = [];
  const byId = new Map<string, Record<string, unknown>>();

  for (const item of raw ?? []) {
    if (typeof item !== "object" || item === null || Array.isArray(item)) {
      errors.push("componente que no es objeto");
      continue;
    }
    const cid = item.id;
    const comp = item.component;
    const props: Record<string, unknown> =
      item.props && typeof item.props === "object"
        ? item.props
        : Object.fromEntries(
            Object.entries(item).filter(([k]) => k !== "id" && k !== "component")
          );

    if (!cid || !comp || typeof cid !== "string" || typeof comp !== "string") {
      errors.push(`componente sin id/component: ${JSON.stringify(item).slice(0, 80)}`);
      continue;
    }
    const compSpec = catalog.components[comp];
    if (!compSpec) {
      errors.push(`'${comp}' no existe en el catálogo (id=${cid})`);
      dropped.push(cid);
      continue;
    }

    const spec = compSpec.props;
    const clean: Record<string, unknown> = {};
    for (const [pname, pspec] of Object.entries(spec)) {
      if (pname in props && props[pname] !== null && props[pname] !== undefined) {
        clean[pname] = coerceProp(comp, cid, pname, pspec, props[pname], errors);
      } else if ("default" in pspec) {
        clean[pname] = pspec.default;
      } else if (pspec.required) {
        errors.push(`${cid}(${comp}) falta prop requerida '${pname}'`);
      }
    }
    for (const extra of Object.keys(props)) {
      if (!(extra in spec)) {
        errors.push(`${cid}(${comp}) prop desconocida '${extra}' (ignorada)`);
      }
    }
    byId.set(cid, { id: cid, component: comp, ...clean });
  }

  // Poda de referencias rotas: children/child que apuntan a ids inexistentes.
  for (const [cid, node] of [...byId.entries()]) {
    if (Array.isArray(node.children)) {
      const alive = (node.children as unknown[]).filter(
        (c) => typeof c === "string" && byId.has(c)
      );
      if (alive.length !== (node.children as unknown[]).length) {
        errors.push(`${cid}: referencias rotas podadas en children`);
      }
      node.children = alive;
    }
    if ("child" in node && !byId.has(node.child as string)) {
      errors.push(`${cid}: child '${String(node.child)}' no existe`);
      delete node.child;
      dropped.push(cid);
      byId.delete(cid);
    }
  }

  return { components: [...byId.values()], errors, dropped };
}

export function compilePlan(
  plan: UiPlan,
  surfaceId: string,
  firstRender: boolean,
  catalog: CatalogDocument
): CompileResult {
  const { components, errors, dropped } = validateComponents(plan.components, catalog);
  const ids = new Set(components.map((c) => c.id as string));
  const root = plan.root || (components.length ? (components[0].id as string) : undefined);

  if (!components.length || !root || !ids.has(root)) {
    errors.push("plan sin root renderizable");
    return { ok: false, messages: [], components: [], data: {}, errors, dropped };
  }

  // Solo un ActionButton puede cerrar el ciclo hacia el agente: sin él la
  // persona no tiene forma de confirmar nada (ver catalog.py — OptionList,
  // Slider y TextField ya no tienen prop "action"). Rechazarlo aquí es lo que
  // fuerza el intento de reparación y, si el modelo insiste, el fallback.
  if (!components.some((c) => c.component === "ActionButton")) {
    errors.push(
      "el plan no incluye ningún ActionButton — toda pantalla necesita " +
        "un botón de confirmación explícito, ningún otro control puede cerrar el ciclo"
    );
    return { ok: false, messages: [], components: [], data: {}, errors, dropped };
  }

  // A2UI v0.9: el root es el primer componente de la lista.
  components.sort((a, b) => (a.id === root ? 0 : 1) - (b.id === root ? 0 : 1));

  const messages: Record<string, unknown>[] = [];
  if (firstRender) messages.push(createSurface(surfaceId, catalog.catalogId));
  const data = plan.data && typeof plan.data === "object" ? plan.data : {};
  if (Object.keys(data).length) messages.push(updateDataModel(surfaceId, "/", data));
  messages.push(updateComponents(surfaceId, components));

  return { ok: true, messages, components, data, errors, dropped };
}

/** Superficie mínima garantizada: la demo nunca se queda en blanco. */
export function fallbackPlan(title: string, body: string, retryAction = "reintentar"): UiPlan {
  return {
    title,
    summary: body,
    root: "root",
    data: {},
    components: [
      { id: "root", component: "Column", props: { children: ["t", "b", "cta"], gap: "md" } },
      { id: "t", component: "Text", props: { text: title, variant: "h2" } },
      { id: "b", component: "Text", props: { text: body, variant: "body" } },
      {
        id: "cta",
        component: "ActionButton",
        props: {
          text: "Reintentar",
          variant: "secondary",
          action: { event: { name: retryAction, params: {} } },
        },
      },
    ],
  };
}
